core/commonMain/src/kotlinx/serialization/encoding/ReorderingCompositeEncoder.kt

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright 2017-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.encoding
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.modules.*
+
+
+/**
+ * Encodes composite elements in a specific order managed by [mapElementIndex].
+ *
+ * This encoder will replicate the behavior of a standard encoding, but calling the `encode*Element` methods in
+ * the order defined by [mapElementIndex]. It first buffers each `encode*Element` calls in an array following
+ * the given indexes using [mapElementIndex], then when [endStructure] is called, it encodes the buffered calls
+ * in the expected order by replaying the previous calls on a concrete [CompositeEncoder] provided
+ * by [beginStructure].
+ *
+ * @param structureDescriptor descriptor of the structure being encoded
+ * @param beginStructure provides the [CompositeEncoder] to be used to encode the given descriptor's elements in the expected order.
+ * @param mapElementIndex maps the element index to the new reordered index (zero-based). If this mapper provides the same index for multiple elements, only the last one will be encoded as the previous ones will be overridden.
+ */
+@ExperimentalSerializationApi
+internal class ReorderingCompositeEncoder(
```

Comment:
If you leave this internal it is not useful. You probably meant for it to be public.

Reply:
This is just in draft, for sure it's not perfect as is was meant to be discussed and have something concrete as a starting point 🙂 let's focus on the content first and then discuss about the public API, I hope it's good for you!